crypto/bio/bio_lib.c

```diff
@@ -94,12 +94,21 @@ int BIO_set(BIO *bio, BIO_METHOD *method)
     bio->num_read = 0L;
     bio->num_write = 0L;
     CRYPTO_new_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
-    if (method->create != NULL)
+
+    bio->lock = CRYPTO_THREAD_lock_new();
+    if (bio->lock == NULL) {
+        BIO_free(bio);
```

Comment:
This will cause double free of 'bio'.
I think doing CRYPTO_free_ex_data() here and let callee do OPENSSL_free (bio) is easier.